File: tools/csharp-roslyn-host/Rules/CodeQuality/MissingAccessModifier.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace TrueCourse.RoslynHost;

/// <summary>
/// A type or type member that declares no explicit access modifier, leaving its
/// accessibility implicit (internal for top-level types, private for members). Making
/// the modifier explicit removes the "is this public?" ambiguity at the declaration.
/// We flag only declarations where an access modifier is *legal and meaningful* —
/// interface members, explicit interface implementations, static constructors,
/// operators, partial members (covered by a separate rule), and enum members are
/// excluded. SA1400.
/// </summary>
internal sealed class MissingAccessModifier : ISemanticRule
{
    public string RuleKey => "code-quality/deterministic/missing-access-modifier";

    private static readonly string[] AccessModifiers = { "public", "private", "protected", "internal" };

    public IEnumerable<Violation> Analyze(SemanticModel model, SyntaxTree tree)
    {
        foreach (var node in tree.GetRoot().DescendantNodes())
        {
            if (node is not MemberDeclarationSyntax member) continue;
            if (!IsAccessModifierApplicable(member)) continue;
            if (HasExplicitAccessModifier(member.Modifiers)) continue;
            // `partial` declarations are governed by partial-element-missing-access-modifier.
            if (member.Modifiers.Any(m => m.IsKind(SyntaxKind.PartialKeyword))) continue;

            var (loc, label) = Target(member);
            if (loc is null) continue;
            var pos = loc.GetLineSpan().StartLinePosition;
            yield return new Violation(
                RuleKey, tree.FilePath, pos.Line + 1, pos.Character + 1,
                $"{label} declares no explicit access modifier; add one to make its accessibility unambiguous.");
        }
    }

    private static bool HasExplicitAccessModifier(SyntaxTokenList modifiers) =>
        modifiers.Any(m => AccessModifiers.Contains(m.Text));

    private static bool IsAccessModifierApplicable(MemberDeclarationSyntax member)
    {
        // Members of an interface have implicit public access that cannot be changed.
        if (member.Parent is InterfaceDeclarationSyntax) return false;
        // Members of an enum (the enum constants) take no access modifier.
        if (member.Parent is EnumDeclarationSyntax) return false;

        switch (member)
        {
            case ClassDeclarationSyntax:
            case StructDeclarationSyntax:
            case InterfaceDeclarationSyntax:
            case RecordDeclarationSyntax:
            case EnumDeclarationSyntax:
            case DelegateDeclarationSyntax:
            case FieldDeclarationSyntax:
            case PropertyDeclarationSyntax:
            case EventDeclarationSyntax:
            case EventFieldDeclarationSyntax:
            case IndexerDeclarationSyntax:
                return true;

            case MethodDeclarationSyntax method:
                // Explicit interface implementations take no access modifier.
                return method.ExplicitInterfaceSpecifier is null;

            case ConstructorDeclarationSyntax ctor:
                // A static constructor never takes an access modifier.
                return !ctor.Modifiers.Any(m => m.IsKind(SyntaxKind.StaticKeyword));

            // Destructors, operators, conversion operators take no access modifier.
            default:
                return false;
        }
    }

    private static (Location?, string) Target(MemberDeclarationSyntax member) => member switch
    {
        BaseTypeDeclarationSyntax t => (t.Identifier.GetLocation(), $"Type '{t.Identifier.ValueText}'"),
        DelegateDeclarationSyntax d => (d.Identifier.GetLocation(), $"Delegate '{d.Identifier.ValueText}'"),
        MethodDeclarationSyntax m => (m.Identifier.GetLocation(), $"Method '{m.Identifier.ValueText}'"),
        PropertyDeclarationSyntax p => (p.Identifier.GetLocation(), $"Property '{p.Identifier.ValueText}'"),
        EventDeclarationSyntax e => (e.Identifier.GetLocation(), $"Event '{e.Identifier.ValueText}'"),
        IndexerDeclarationSyntax ix => (ix.ThisKeyword.GetLocation(), "Indexer"),
        ConstructorDeclarationSyntax c => (c.Identifier.GetLocation(), $"Constructor '{c.Identifier.ValueText}'"),
        FieldDeclarationSyntax f when f.Declaration.Variables.Count > 0
            => (f.Declaration.Variables[0].Identifier.GetLocation(), $"Field '{f.Declaration.Variables[0].Identifier.ValueText}'"),
        EventFieldDeclarationSyntax ef when ef.Declaration.Variables.Count > 0
            => (ef.Declaration.Variables[0].Identifier.GetLocation(), $"Event '{ef.Declaration.Variables[0].Identifier.ValueText}'"),
        _ => (null, ""),
    };
}

File: packages/analyzer/src/rules/code-quality/visitors/csharp/missing-access-modifier.ts
import type { Node as SyntaxNode } from 'web-tree-sitter'
import type { CodeRuleVisitor } from '../../../types.js'
import { makeViolation } from '../../../types.js'

const ACCESS_MODIFIERS = new Set(['public', 'private', 'protected', 'internal'])

const MEMBER_TYPES = [
  'class_declaration', 'struct_declaration', 'interface_declaration', 'record_declaration',
  'enum_declaration', 'delegate_declaration', 'field_declaration', 'property_declaration',
  'event_declaration', 'event_field_declaration', 'indexer_declaration',
  'method_declaration', 'constructor_declaration',
]

function modifierTexts(node: SyntaxNode): string[] {
  return node.children.filter((c) => c?.type === 'modifier').map((c) => c!.text)
}

function declNameNode(node: SyntaxNode): SyntaxNode {
  if (node.type === 'field_declaration' || node.type === 'event_field_declaration') {
    const decl = node.namedChildren.find((c) => c?.type === 'variable_declaration')
    const declarator = decl?.namedChildren.find((c) => c?.type === 'variable_declarator')
    return declarator?.childForFieldName('name') ?? node
  }
  return node.childForFieldName('name') ?? node
}

/**
 * A type or member that declares no explicit access modifier, leaving its
 * accessibility to language defaults (internal for top-level types, private for
 * members) — easy to misread and easy to widen by accident. Interface members
 * (implicitly public), enum constants, explicit interface implementations and
 * static constructors take no access modifier and are never flagged; `partial`
 * declarations are covered by partial-element-missing-access-modifier.
 */
export const csharpMissingAccessModifierVisitor: CodeRuleVisitor = {
  ruleKey: 'code-quality/deterministic/missing-access-modifier',
  languages: ['csharp'],
  nodeTypes: MEMBER_TYPES,
  visit(node, filePath, sourceCode) {
    const mods = modifierTexts(node)
    if (mods.some((m) => ACCESS_MODIFIERS.has(m))) return null
    if (mods.includes('partial')) return null

    // Members of an interface are implicitly public.
    if (node.parent?.parent?.type === 'interface_declaration') return null

    if (node.type === 'method_declaration') {
      // Explicit interface implementations (`void IFoo.Bar()`) take no modifier.
      const nm = node.childForFieldName('name')
      if (nm?.type === 'qualified_name' || (nm?.text ?? '').includes('.')) return null
      if (node.namedChildren.some((c) => c?.type === 'explicit_interface_specifier')) return null
    }
    if (node.type === 'constructor_declaration' && mods.includes('static')) return null

    const target = declNameNode(node)
    return makeViolation(
      this.ruleKey, target, filePath, 'low',
      'Missing access modifier',
      'Declaration has no explicit access modifier; add one to make its accessibility unambiguous.',
      sourceCode,
      'Add an explicit access modifier (public/internal/protected/private).',
    )
  },
}
